src/main/scala/nodes/learning/KernelRidgeRegression.scala

```diff
@@ -0,0 +1,240 @@
+      }
+    }.setName("model").cache()
+    model.count
+
+    // We also keep a local copy of the model to use computing residual
+    // NOTE: The last block might have fewer than blockSize features
+    // but we will correct this inside the solve
+    val wLocals = lambdas.map { l =>
+      (0 until numBlocks).map { x =>
+        DenseMatrix.zeros[Double](blockSize, nClasses)
+      }.toArray
+    }
+    val blockShuffler = blockPermuter.map(seed => new Random(seed))
+
+    (0 until numEpochs).foreach { pass =>
+      val inOrder = (0 until numBlocks).toIndexedSeq
+      val blockOrder = blockShuffler.map(rnd => rnd.shuffle(inOrder)).getOrElse(inOrder)
+
+      blockOrder.foreach { block =>
+        val blockBegin = System.nanoTime
+        // Use the kernel block generator to get kernel matrix for these column blocks
+        val blockIdxs = (blockSize * block) until (math.min(nTrain, (block + 1) * blockSize))
+        val blockIdxsSeq = blockIdxs.toArray
+        val blockIdxsBC = labelsMat.context.broadcast(blockIdxsSeq)
+
+        val kernelBlockMat = trainKernelMat(blockIdxsSeq).cache()
+        val kernelBlockBlockMat = DenseMatrix.vertcat(trainKernelMat(blockIdxsSeq, blockIdxsSeq).collect():_*) 
+
+        // Build up the residual
+        val treeBranchingFactor = labels.context.getConf.getInt(
+          "spark.mlmatrix.treeBranchingFactor", 2).toInt
+        val depth = max(math.ceil(math.log(kernelBlockMat.partitions.size) /
+          math.log(treeBranchingFactor)).toInt, 1)
+
+        // Compute K_B^T * W as its easy to do this.
+        // After this we will subtract out K_BB^T * W_B
+        // b x k
+        val residuals = MLMatrixUtils.treeReduce(kernelBlockMat.zip(model).map { x =>
+          x._2.map { y =>
+            // this is a b * n1 times n1 * k
+            x._1.t * y
+          }
+        }, addMatrixArrays, depth=depth)
+
+        // This is b*k
+        val y_bb = labelsRPM(blockIdxs, ::).collect()
+
+        // This is a tuple of (oldBlockBC, newBlockBC)
+        val wBlockBCs = (0 until lambdas.length).map { l =>
+          // This is b*k
+          val wBlockOld = if (pass == 0) {
+            DenseMatrix.zeros[Double](blockIdxs.size, nClasses)
+          } else {
+            wLocals(l)(block)
+          }
+          val lhs = kernelBlockBlockMat +
+            DenseMatrix.eye[Double](kernelBlockBlockMat.rows) * lambdas(l)
+          // Subtract out K_bb * W_bb from residual
+          val rhs = y_bb - (residuals(l) - kernelBlockBlockMat.t * wBlockOld)
+
+          val wBlockNew = lhs \ rhs
+          wLocals(l)(block) = wBlockNew
+          val wBlockOldBC = labels.context.broadcast(wBlockOld)
+          val wBlockNewBC = labels.context.broadcast(wBlockNew)
+          (wBlockOldBC, wBlockNewBC)
+        }
+
+        var newModel = updateModel(
+          model, wBlockBCs.map(_._2), blockIdxsBC, preFixLengthsBC).cache()
+        // This is to truncate the lineage every 50 blocks
```

Comment:
Is there a reference to a `SPARK-` issue we can put here, or is it just the case that RDDs with long lineage suck.